Basic backwards compatibility

liamsi opened the pull request:
To achieve basic backwards compatibility, we continue reading the passed buffer even if all fields were already read. We only check if what follows is valid amino encoding.

resolves #143 

this also includes changes of #153 which 
resolves #145

liamsi committed:
WIP: working backwards compatibility, needs tests

liamsi committed:
added a simple test

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
-		var typ = Typ3(0x00)
+		// when encountering an StructTerm typ3 byte.
+
+		var typ Typ3
```
This and the comments above belong below the for-loop

liamsi replied:
following 
> Then, line 716~723 below is unneeded

I deleted the part outside the for-loop and with it this comment(s).

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+
+		var typ Typ3
+		// we have read all, or do we? consume the rest
+		for len(bz) > 2 && bz[0]&0xF8 != 0 && Typ3(bz[0]) != Typ3_StructTerm {
```
I think we just want a for loop with no condition here, and below in `decodeFieldNumberAndTyp3(bz)` break (without sliding) if typ3 is Typ3_StructTerm.  

Then, line 716~723 below is unneeded, as the unconditional for-loop will return an error if after sliding bz is insufficient in length.  Ergo no need for len(bz) > 2 in the if-condition.

liamsi replied:
I added a small test to see if this works.

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+		var typ Typ3
+		// we have read all, or do we? consume the rest
+		for len(bz) > 2 && bz[0]&0xF8 != 0 && Typ3(bz[0]) != Typ3_StructTerm {
+			fnum, typ3, _n, _ := decodeFieldNumberAndTyp3(bz)
```
But we'd have to check the error instead of assigning to _.  
In general it's a bad idea to ignore errors, usually it's wrong.

liamsi replied:
Good catch! The whole purpose of this line is to catch potential errors ...

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+		// we have read all, or do we? consume the rest
+		for len(bz) > 2 && bz[0]&0xF8 != 0 && Typ3(bz[0]) != Typ3_StructTerm {
+			fnum, typ3, _n, _ := decodeFieldNumberAndTyp3(bz)
+			if slide(&bz, &n, _n) && err != nil {
```
This err is the wrong err to check, as it wasn't assigned from `decodeFieldNumberAndTyp3`.

liamsi replied:
fixed

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+			if slide(&bz, &n, _n) && err != nil {
+				return
+			}
+			if fnum > uint32(len(info.Fields)) {
```
This conditional should rather be a sanity check, that fnum > `greatest field number seen`.  Instead of len(info.Fields) (which isn't forward-compatible once we support custom field tag numbers), above line 651 we should keep track of `var lastFieldNum` which we keep incrementing in both for loops (651 and 703), and we should return an error from either for-loop if each newly encountered field doesn't have a greater field number than `lastFieldNum`.

As is, this conditional block will just be ignored if fnum doesn't increment, but lets be more restrictive in our decoding.

liamsi replied:
Perfect 💯Thanks a lot for the thorough review and the suggestions!

jaekwon commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+	return
+}
+
+// read everything without doing anything with it
```
Full sentence, so capital r and ends with a period.

jaekwon reviewed (changes requested):
Left comments in the file diff view

rigelrozanski commented on binary-decode.go:
```diff
@@ -677,12 +678,12 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 			// so we will need to remove this sanity check.
 			if field.BinFieldNum != fieldNum {
-				err = errors.New(fmt.Sprintf("Expected field number %v, got %v", field.BinFieldNum, fieldNum))
+				err = errors.New(fmt.Sprintf("expected field number %v, got %v", field.BinFieldNum, fieldNum))
```
use `fmt.Errorf` throughout instead of `errors.New(fmt.Sprintf(`

liamsi replied:
done; with all other occurrences in this file

rigelrozanski commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+}
+
+func consumeStruct(bz []byte) (n int, err error) {
+	var _n, typ = int(0), Typ3(0x00)
```
what's this `_n` variable I keep seeing - as an outsider I'm left in the dark also typically don't want to use underscores in names, but depending on what `_n` is it may be appropriate

liamsi replied:
for the sake of consistency this follows: https://github.com/tendermint/go-amino/blob/ecf42f8907e063f630fcce7605c61252db93c355/reflect.go#L42-L51

I'm happy to change this but would do this everywhere in the code (and not only here) in a separate issue.

rigelrozanski replied:
yup sounds good - I'm the outsider here hehe

rigelrozanski commented on binary-decode.go:
```diff
@@ -695,18 +696,158 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 			return
 		}
+		if term {
+			break FOR_LOOP
```
line jumps makes for spaghetti code, avoid at all costs - in this situation you can just use `break` 
```golang
if term {
	break
}
```
and then delete the like with `FOR_LOOP:`

liamsi replied:
done (replace `break` with `return` though, b/c that's what happens after the for loop)

liamsi committed:
added another test and review comments

jaekwon commented on binary-decode.go:
```diff
@@ -630,9 +631,12 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 	}
 	_n := 0 // nolint: ineffassign
 
+	// Keep incrementing the last fieldnum. Later, when we support custom fieldtag numbers, we need to store the last
```
should be declared in the case statement in the switch statement?

liamsi replied:
Thx, done.

jaekwon commented on binary-decode.go:
```diff
@@ -648,6 +652,7 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 	default:
 		// Read each field.
 		for _, field := range info.Fields {
+			lastFieldNum++
```
It's semantically awkward to increment it here in the beginning... Instead I'd increment at the end of the for-loop.

liamsi replied:
See [below](https://github.com/tendermint/go-amino/pull/155/files#r191466725). I don't increment now but set `lastFieldNum = fieldnum` instead (not at the end but right after the check)

jaekwon commented on binary-decode.go:
```diff
@@ -670,19 +675,25 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
 				// Probably a StructTerm.
 				break
 			}
+			if fieldNum < lastFieldNum {
```
like, semantically it should be `if fieldNum <= lastFieldNum`, because it shouldn't be lastFieldNum either.

jaekwon commented on binary-decode.go:
```diff
@@ -693,20 +704,163 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+		var typ3 Typ3
+		// we have read all, or do we? consume the rest
+		for {
+			lastFieldNum++
```
This is especially awkward here because...

jaekwon commented on binary-decode.go:
```diff
@@ -693,20 +704,163 @@ func (cdc *Codec) decodeReflectBinaryStruct(bz []byte, info *TypeInfo, rv reflec
+			if slide(&bz, &n, _n) && err != nil {
+				return
+			}
+			if fnum < lastFieldNum {
```
... it should be `if fnum <= lastFieldNum { ... } else { lastFieldNum = fnum }`.  The `lastFieldNum++` above is incorrect.

liamsi replied:
Then, in the 1st check above (in the 1st for-loop) should also not just increment (`lastFieldNum++`) but set to `lastFieldNum = fnum` instead. Or is there any reason why this should be different only here? 

If we increment like before (in the beginning of the loop and stick to `<`) it is indeed semantically awkward (totally agree) but not incorrect. At least, I tried constructing a test where decoding fails but couldn't (only by manually modifying the bytes to have fieldnums which are not just start with zero and then "last + 1"; in other words with "gaps" in between the numbering). 

Always setting `lastFieldNum = fnum` will also work if we add manual field numbering support. Incrementing only works because the numbering is implicitly given by the field order.

jaekwon reviewed (changes requested):
Added more comments in file change view

liamsi committed:
review comments: lastFieldNum now checks that field nums are
monotonically increasing (strictly) instead of in +1 steps (allow 'gaps')
modified test to be slightly more complex (added time)

jaekwon committed:
Merge branch 'develop' into backwards_compat
